File: api/screenshot.py
# Imports

## Flask Management
from flask import Flask, request, Response
from flask_compress import Compress
from flask_cors import CORS

## Initialization
app = Flask(__name__) # Flask
Compress(app) # Compressing the responses
CORS(app) # Enabling CORS globally

## Other Imports
from time import time
from traceback import print_exc
from constants import makeResponse, to_bool
from driver.chrome import Chrome
from uuid import uuid4

## Variable Initialization
LOCAL_CACHE = {} # store variable
CACHE_DURATION = 3 # in sec.

CHROME = None
PID = None

# Route Defining
@app.route('/', defaults={'path': ''})
@app.route('/<path:path>')
def hello(path):
    try:
        ## if the name of the file is in the following schema: [parameter].py
        # parameter = path[path.rfind("/") + 1:]

        # Cache Management

        ## Cache Key Defining
        _values_dict = request.values.to_dict()
        _values_dict.pop("_invalidate_cache", None)
        cache_key = str(_values_dict) + str(request.method)

        ## Cache Invalidation Management
        if "_invalidate_cache" in request.values:
            LOCAL_CACHE.pop(cache_key, None)
            return "Ok"

        if "preload" in request.values:
            global CHROME
            global PID
            if CHROME is not None:
                CHROME.close()
                CHROME.quit()
            CHROME = Chrome(width=request.values.get("width", 1920, int), height=request.values.get("height", 1080, int))
            PID = uuid4()
            return makeResponse({"pid": PID, "width": CHROME.width, "height": CHROME.height})

        if "loaded" in request.values:
            if CHROME is not None:
                return makeResponse({"pid": PID, "width": CHROME.width, "height": CHROME.height})
            else:
                return makeResponse({"pid": None, "width": None, "height": None})
        
        ## Cache Lookup
        try:
            if cache_key in LOCAL_CACHE:
                cache_duration = time() - LOCAL_CACHE[cache_key]["timestamp"]
                if cache_duration > CACHE_DURATION:
                    LOCAL_CACHE.pop(cache_key, None)
                else:
                    return Response(LOCAL_CACHE[cache_key]["data"], content_type=LOCAL_CACHE[cache_key]["type"], headers={"X-ANISE-CACHE": "HIT"})
        except:
            print_exc()
            print("[CACHE] An error occured while sending back the cached data")
            print("[FAILURE_RECOVERING] Processing the request as if nothing was cached")

        # Processing and Computation
        base64 = to_bool(request.values.get("base64", False))
        result = CHROME.screenshot(request.values.get("url", None), width=request.values.get("width", None, int), height=request.values.get("height", None, int), base64=base64) # the result should in this variable
        
        # Caching and Response
        LOCAL_CACHE[cache_key] = {"timestamp": time(), "data": result, "type": "application/json" if base64 else "image/png"}
        if base64:
            return makeResponse({"pid": PID, "width": CHROME.width, "height": CHROME.height, "result": result}, cache_hit=False)
        return Response(result, content_type="image/png", headers={"X-ANISE-CACHE": "MISS"})
    except:
        print_exc()
        print("[ERROR] An unknown error occured on the server and nothing could handle it. Sending back SERVER_ERROR (Status Code: 500)")
        return makeResponse({"message": "An error occured on the server"}, error="SERVER_ERROR", code=500)

File: api/screenshot.ts
import { VercelRequest, VercelResponse } from '@vercel/node';
import { Chrome, Size } from '../driver/chrome';
import { Response, sendResponse } from "../utils/response";
import { v4 as uuidv4 } from 'uuid';
import toBool from '../utils/boolean';
import log from "../utils/log";

let CHROME: Chrome = null;
let PID: string = null;

export default (request: VercelRequest, vercelResponse: VercelResponse) => {
    let width: number = null
    if (request.query.width) {
        width = Number(request.query.width)
    }
    let height: number = null
    if (request.query.height) {
        height = Number(request.query.height)
    }
    if (request.query.loaded && toBool(request.query.loaded)) {
        log("Requesting for the current loaded Chromium instance metadata", "/screenshot endpoint")
        if (CHROME) {
            log("Chromium IS loaded", "/screenshot endpoint")
            var response = new Response({
                "pid": PID,
                "width": CHROME.size.width,
                "height": CHROME.size.height
            })
        } else {
            log("Chromium is NOT loaded", "/screenshot endpoint")
            var response = new Response({
                "pid": null,
                "width": null,
                "height": null
            })
        }
        sendResponse(vercelResponse, response)
    } else if (request.query.preload && toBool(request.query.preload)) {
        log("Requesting to preload Chromium", "/screenshot endpoint")
        if (CHROME) {
            log("Closing existing Chromium instance", "/screenshot endpoint")
            CHROME.close()
        }
        CHROME = new Chrome()
        CHROME.init(width, height).then(() => {
            PID = uuidv4()
            let response = new Response({
                "pid": PID,
                "width": CHROME.size.width,
                "height": CHROME.size.height
            })
            sendResponse(vercelResponse, response)
        })
    } else {
        if (!CHROME) {
            CHROME = new Chrome();
            PID = uuidv4();
        }
        CHROME.init(width, height).then(() => {
            let url: string = null;
            if (request.query.url) {
                url = request.query.url.toString()
            }
            let base64: boolean = false;
            if (request.query.base64) {
                base64 = toBool(request.query.url)
            }
            let full: boolean = false;
            if (request.query.url) {
                full = toBool(request.query.full)
            }
            CHROME.screenshot(url, base64, full).then((result) => {
                if (base64) {
                    let response = new Response({
                        "pid": PID,
                        "width": CHROME.size.width,
                        "height": CHROME.size.height,
                        "result": result
                    })
                    sendResponse(vercelResponse, response)
                } else {
                    vercelResponse.setHeader("X-ANISE-CACHE", "MISS")
                    vercelResponse.setHeader("Server", "Anise")
                    vercelResponse.setHeader("Content-Type", "image/png")
                    vercelResponse.status(200).send(result)
                }
            })
        })
    }
};